File: android/src/main/kotlin/co/tamara/sdk/TamaraPlugin.kt
package co.tamara.sdk

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.widget.Toast
import com.google.gson.Gson
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry

class TamaraPlugin: FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware, PluginRegistry.ActivityResultListener, Tamara {
    private val CHANNEL = "co.tamara.sdk"
    private lateinit var _result: MethodChannel.Result
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
        context = binding.applicationContext
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initSdk" -> {
                val authToken : String? = call.argument("authToken")
                val apiUrl : String? = call.argument("apiUrl")
                val notificationWebHook : String? = call.argument("notificationWebHook")
                val publishKey : String? = call.argument("publishKey")
                val notificationToken : String? = call.argument("notificationToken")
                val isSanbox: Boolean? = call.argument("isSandbox")
                init(authToken = authToken!!, apiUrl = apiUrl!!, notificationWebHook = notificationWebHook!!,
                    publishKey = publishKey!!, notificationToken = notificationToken!!, isSanbox = isSanbox!!)
            }
            "startPayment" -> {
                val checkoutUrl : String? = call.argument("checkoutUrl")
                val successCallbackUrl : String? = call.argument("successCallbackUrl")
                val failureCallbackUrl : String? = call.argument("failureCallbackUrl")
                val cancelCallbackUrl : String? = call.argument("cancelCallbackUrl")
                paymentCheckout(checkoutUrl!!, successCallbackUrl!!, failureCallbackUrl!!, cancelCallbackUrl!!)
            }
            "orderDetail" -> {
                _result = result
                val orderId : String? = call.argument("orderId")
                getOrderDetail(orderId!!)
            }
            "capturePayment" -> {
                _result = result
                val jsonData : String? = call.argument("capturePayment")
                getCapturePayment(jsonData!!)
            }
            "createOrder" -> {
                val orderReferenceId : String? = call.argument("orderReferenceId")
                val description : String? = call.argument("description")
                createOrder(orderReferenceId+"", description+"")
            }
            "setCountry" -> {
                val countryCode : String? = call.argument("countryCode")
                val currency : String? = call.argument("currency")
                setCountry(countryCode!!, currency!!)
            }
            "setPaymentType" -> {
                val paymentType : String? = call.argument("paymentType")
                setPaymentType(paymentType!!)
            }
            "setInstalments" -> {
                val instalments : Int? = call.argument("instalments")
                setInstalments(instalments!!)
            }
            "setCustomerInfo" -> {
                val firstName : String? = call.argument("firstName")
                val lastName : String? = call.argument("lastName")
                val phoneNumber : String? = call.argument("phoneNumber")
                val email : String? = call.argument("email")
                val isFirstOrder : Boolean? = call.argument("isFirstOrder")
                setCustomerInfo(firstName!!, lastName!!, phoneNumber!!, email!!, isFirstOrder!!)
            }
            "addItem" -> {
                val name : String? = call.argument("name")
                val referenceId : String? = call.argument("referenceId")
                val sku : String? = call.argument("sku")
                val type : String? = call.argument("type")
                val unitPrice : Double? = call.argument("unitPrice")
                val tax : Double? = call.argument("tax")
                val discount : Double? = call.argument("discount")
                val quantity : Int? = call.argument("quantity")
                addItem(name!!, referenceId!!, sku!!, type!!, unitPrice!!, tax!!, discount!!, quantity!!)
            }
            "setShippingAddress" -> {
                val firstName : String? = call.argument("firstName")
                val lastName : String? = call.argument("lastName")
                val phoneNumber : String? = call.argument("phoneNumber")
                val addressLine1 : String? = call.argument("addressLine1")
                val addressLine2 : String? = call.argument("addressLine2")
                val country : String? = call.argument("country")
                val region : String? = call.argument("region")
                val city : String? = call.argument("city")
                setShippingAddress(firstName!!, lastName!!, phoneNumber!!, addressLine1!!, addressLine2!!, country!!, region!!, city!!)
            }
            "setBillingAddress" -> {
                val firstName : String? = call.argument("firstName")
                val lastName : String? = call.argument("lastName")
                val phoneNumber : String? = call.argument("phoneNumber")
                val addressLine1 : String? = call.argument("addressLine1")
                val addressLine2 : String? = call.argument("addressLine2")
                val country : String? = call.argument("country")
                val region : String? = call.argument("region")
                val city : String? = call.argument("city")
                setBillingAddress(firstName!!, lastName!!, phoneNumber!!, addressLine1!!, addressLine2!!, country!!, region!!, city!!)
            }
            "setCurrency" -> {
                val newCurrency : String? = call.argument("newCurrency")
                setCurrency(newCurrency!!)
            }
            "setShippingAmount" -> {
                val amount : Double? = call.argument("amount")
                setShippingAmount(amount!!)
            }
            "setDiscount" -> {
                val amount : Double? = call.argument("amount")
                val name : String? = call.argument("name")
                setDiscount(amount!!, name!!)
            }
            "paymentOrder" -> {
                _result = result
                paymentOrder()
            }
            "refunds" -> {
                _result = result
                val orderId : String? = call.argument("orderId")
                val jsonData : String? = call.argument("refunds")
                refunds(orderId!!, jsonData!!)
            }
            "cancelOrder" -> {
                _result = result
                val orderId : String? = call.argument("orderId")
                val jsonData : String? = call.argument("cancelOrder")
                cancelOrder(orderId!!, jsonData!!)
            }
            "updateOrderReference" -> {
                _result = result
                val orderId : String? = call.argument("orderId")
                val orderReferenceId : String? = call.argument("orderReferenceId")
                updateOrderReference(orderId!!, orderReferenceId!!)
            }
            "renderCartPage" -> {
                _result = result
                val language : String? = call.argument("language")
                val country : String? = call.argument("country")
                val publicKey : String? = call.argument("publicKey")
                val amount : Double? = call.argument("amount")
                renderCartPage(language!!, country!!, publicKey!!, amount!!)
            }
            "renderProduct" -> {
                _result = result
                val language : String? = call.argument("language")
                val country : String? = call.argument("country")
                val publicKey : String? = call.argument("publicKey")
                val amount : Double? = call.argument("amount")
                renderProduct(language!!, country!!, publicKey!!, amount!!)
            }
            "authoriseOrder" -> {
                _result = result
                val orderId : String? = call.argument("orderId")
                authoriseOrder(orderId!!)
            }
            "clearItem" -> {
                clearItem()
            }
            else -> {
            }
        }
    }

    override fun init(
        authToken: String,
        apiUrl: String,
        notificationWebHook: String,
        publishKey: String,
        notificationToken: String,
        isSanbox: Boolean
    ) {
        TamaraPayment.initialize(authToken, apiUrl, notificationWebHook, publishKey, notificationToken, isSanbox)
    }

    override fun paymentCheckout(
        checkOutUrl: String,
        successCallbackUrl: String,
        failureCallbackUrl: String,
        cancelCallbackUrl: String
    ) {
        TamaraPayment.startPayment(activity!!, checkOutUrl, successCallbackUrl, failureCallbackUrl, cancelCallbackUrl)
    }

    override fun getOrderDetail(orderId: String) {
        TamaraPayment.getOrderDetail(activity!!, orderId)
    }

    override fun getCapturePayment(jsonData: String) {
        TamaraPayment.getCapturePayment(activity!!, jsonData)
    }

    override fun createOrder(orderReferenceId: String, description: String) {
        TamaraPayment.createOrder(orderReferenceId, description)
    }

    override fun setCountry(countryCode: String, currency: String) {
        TamaraPayment.setCountry(countryCode, currency)
    }

    override fun setPaymentType(paymentType: String) {
        TamaraPayment.setPaymentType(paymentType)
    }

    override fun setInstalments(instalments: Int) {
        TamaraPayment.setInstalments(instalments)
    }

    override fun setCustomerInfo(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        email: String,
        isFirstOrder: Boolean
    ) {
        TamaraPayment.setCustomerInfo(firstName, lastName, phoneNumber, email, isFirstOrder)
    }

    override fun addItem(
        name: String,
        referenceId: String,
        sku: String,
        type: String,
        unitPrice: Double,
        tax: Double,
        discount: Double,
        quantity: Int
    ) {
        TamaraPayment.addItem(name, referenceId, sku, type, unitPrice, tax, discount, quantity)
    }

    override fun setShippingAddress(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        addressLine1: String,
        addressLine2: String,
        country: String,
        region: String,
        city: String
    ) {
        TamaraPayment.setShippingAddress(firstName, lastName, phoneNumber, addressLine1, addressLine2, country, region, city)
    }

    override fun setBillingAddress(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        addressLine1: String,
        addressLine2: String,
        country: String,
        region: String,
        city: String
    ) {
        TamaraPayment.setBillingAddress(firstName, lastName, phoneNumber, addressLine1, addressLine2, country, region, city)
    }

    override fun setCurrency(newCurrency: String) {
        TamaraPayment.setCurrency(newCurrency)
    }

    override fun setShippingAmount(amount: Double) {
        TamaraPayment.setShippingAmount(amount)
    }

    override fun setDiscount(amount: Double, name: String) {
        TamaraPayment.setDiscount(amount, name)
    }

    override fun paymentOrder() {
        TamaraPayment.startPayment(activity!!)
    }

    override fun refunds(orderId: String, jsonData: String) {
        TamaraPayment.refunds(activity!!, orderId, jsonData)
    }

    override fun cancelOrder(orderId: String, jsonData: String) {
        TamaraPayment.cancelOrder(activity!!, orderId, jsonData)
    }

    override fun updateOrderReference(orderId: String, orderReferenceId: String) {
        TamaraPayment.updateOrderReference(activity!!, orderId, orderReferenceId)
    }

    override fun renderCartPage(
        language: String,
        country: String,
        publicKey: String,
        amount: Double
    ) {
        TamaraPayment.renderWidgetCartPage(activity!!, language, country, publicKey, amount)
    }

    override fun renderProduct(
        language: String,
        country: String,
        publicKey: String,
        amount: Double
    ) {
        TamaraPayment.renderWidgetProduct(activity!!, language, country, publicKey, amount)
    }

    override fun authoriseOrder(orderId: String) {
        TamaraPayment.authoriseOrder(activity!!, orderId)
    }

    override fun clearItem() {
        TamaraPayment.clearItem()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        when(requestCode){
            TamaraPayment.REQUEST_TAMARA_PAYMENT -> {
                if(TamaraPaymentHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraPaymentHelper.getData(data!!)
                    when(result?.status){
                        PaymentResult.STATUS_CANCEL ->{
                            _result.success("Payment canceled")
                            Toast.makeText(activity, "Payment canceled", Toast.LENGTH_LONG).show()
                        }
                        PaymentResult.STATUS_FAILURE -> {
                            _result.success(result.getMessage() ?: "Payment error")
                            Toast.makeText(activity, result.getMessage() ?: "Payment error", Toast.LENGTH_LONG).show()
                        }
                        PaymentResult.STATUS_SUCCESS -> {
                            val checkout = TamaraPaymentHelper.checkOutSession(data);
                            _result.success(Gson().toJson(checkout))
                        }
                    }
                }
            }
            TamaraPayment.REQUEST_TAMARA_ORDER_DETAIL -> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Order Detail error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val orderDetail = TamaraInformationHelper.getDataOrderDetail(data)
                            _result.success(Gson().toJson(orderDetail))
                        }
                    }
                }
            }

            TamaraPayment.REQUEST_TAMARA_CAPTURE_PAYMENT-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Capture Payment error", Toast.LENGTH_LONG).show()
                            _result.success(result.getMessage() ?: "Capture Payment error")
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val capturePayment = TamaraInformationHelper.getDataCapturePayment(data)
                            _result.success(Gson().toJson(capturePayment))
                        }
                    }
                }
            }
            TamaraPayment.REQUEST_TAMARA_REFUNDS-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Refunds error", Toast.LENGTH_LONG).show()
                            _result.success(result.getMessage() ?: "Refunds error")
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val refunds = TamaraInformationHelper.getDataRefunds(data)
                            _result.success(Gson().toJson(refunds))
                        }
                    }
                }
            }
            TamaraPayment.REQUEST_TAMARA_REFERENCE-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Order Reference error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val orderReference = TamaraInformationHelper.getOrderReference(data)
                            _result.success(Gson().toJson(orderReference))
                        }
                    }
                }
            }

            TamaraPayment.REQUEST_TAMARA_CART_PAGE-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Cart page error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val cartPage = TamaraInformationHelper.getCartPage(data)
                            _result.success(Gson().toJson(cartPage))
                        }
                    }
                }
            }

            TamaraPayment.REQUEST_TAMARA_PRODUCT-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Product error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val product = TamaraInformationHelper.getProduct(data)
                            _result.success(Gson().toJson(product))
                        }
                    }
                }
            }

            TamaraPayment.REQUEST_TAMARA_CANCEL-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Cancel order error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val cancelOrder = TamaraInformationHelper.getCancelOrder(data)
                            _result.success(Gson().toJson(cancelOrder))
                        }
                    }
                }
            }
            TamaraPayment.REQUEST_TAMARA_AUTHORISE-> {
                if(TamaraInformationHelper.shouldHandleActivityResult(requestCode, resultCode, data)){
                    var result = TamaraInformationHelper.getData(data!!)
                    when(result?.status){
                        InformationResult.STATUS_FAILURE -> {
                            Toast.makeText(activity, result.getMessage() ?: "Authorise order error", Toast.LENGTH_LONG).show()
                        }
                        InformationResult.STATUS_SUCCESS -> {
                            val authorise = TamaraInformationHelper.getAuthoriseOrder(data)
                            Handler(Looper.getMainLooper()).post {
                                _result.success(Gson().toJson(authorise))
                            }
                        }
                    }
                }
            }
        }

        return false
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        binding.addActivityResultListener(this)
        activity = binding.getActivity()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.getActivity()
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}

File: ios/Classes/TamaraPlugin.swift
import UIKit
import Flutter
import Foundation

public class TamaraPlugin: NSObject, FlutterPlugin {
   public static func register(with registrar: FlutterPluginRegistrar) {
       let channel = FlutterMethodChannel(name: "co.tamara.sdk/ios", binaryMessenger: registrar.messenger())
       let instance = TamaraPlugin()
       registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
            switch call.method {
            case "initSdk":
                guard let args = call.arguments as? [String : Any] else {return}
                let authToken = args["authToken"] as! String
                let apiUrl = args["apiUrl"] as! String
                let notificationWebHook = args["notificationWebHook"] as! String
                let publishKey = args["publishKey"] as! String
                let notificationToken = args["notificationToken"] as! String
                let isSandbox = args["isSandbox"] as! Bool
                self.initialize(token: authToken, apiUrl: apiUrl, pushUrl: notificationWebHook, publishKey: publishKey, notificationToken: notificationToken, isSandbox: isSandbox)
                print("number: 0") // In ra number: 0
            case "createOrder":
                guard let args = call.arguments as? [String : Any] else {return}
                let orderReferenceId = args["orderReferenceId"] as! String
                let description = args["description"] as! String
                self.createOrder(orderReferenceId: orderReferenceId, description: description)
            case "setCountry":
                guard let args = call.arguments as? [String : Any] else {return}
                let countryCode = args["countryCode"] as! String
                let currency = args["currency"] as! String
                self.setCountry(countryCode: countryCode, currency: currency)
            case "setPaymentType" :
                guard let args = call.arguments as? [String : Any] else {return}
                let paymentType = args["paymentType"] as! String
                self.setPaymentType(paymentType: paymentType)
            case "setInstalments" :
                guard let args = call.arguments as? [String : Int] else {return}
                let instalments = args["instalments"] as! Int
                self.setInstalments(instalments: instalments)
            case "setCustomerInfo":
                guard let args = call.arguments as? [String : Any] else {return}
                let firstName = args["firstName"] as! String
                let lastName = args["lastName"] as! String
                let phoneNumber = args["phoneNumber"] as! String
                let email = args["email"] as! String
                let isFirstOrder = args["isFirstOrder"] as! Bool
                self.setCustomerInfo(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, email: email, isFirstOrder: isFirstOrder)
            case "addItem" :
                guard let args = call.arguments as? [String : Any] else {return}
                let name = args["name"] as! String
                let referenceId = args["referenceId"] as! String
                let sku = args["sku"] as! String
                let type = args["type"] as! String
                let unitPrice = args["unitPrice"] as! Double
                let tax = args["tax"] as! Double
                let discount = args["discount"] as! Double
                let quantity = args["quantity"] as! Int
                self.addItem(name: name, referenceId: referenceId, sku: sku, type: type, unitPrice: unitPrice, tax: tax, discount: discount, quantity: quantity)
            case "setShippingAddress" :
                guard let args = call.arguments as? [String : Any] else {return}
                let firstName = args["firstName"] as! String
                let lastName = args["lastName"] as! String
                let phoneNumber = args["phoneNumber"] as! String
                let addressLine1 = args["addressLine1"] as! String
                let addressLine2 = args["addressLine2"] as! String
                let country = args["country"] as! String
                let region = args["region"] as! String
                let city = args["city"] as! String
                self.setShippingAddress(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, addressLine1: addressLine1, addressLine2: addressLine2, country: country, region: region, city: city)
            case "setBillingAddress" :
                guard let args = call.arguments as? [String : Any] else {return}
                let firstName = args["firstName"] as! String
                let lastName = args["lastName"] as! String
                let phoneNumber = args["phoneNumber"] as! String
                let addressLine1 = args["addressLine1"] as! String
                let addressLine2 = args["addressLine2"] as! String
                let country = args["country"] as! String
                let region = args["region"] as! String
                let city = args["city"] as! String
                self.setBillingAddress(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, addressLine1: addressLine1, addressLine2: addressLine2, country: country, region: region, city: city)
            case "setCurrency" :
                guard let args = call.arguments as? [String : Any] else {return}
                let newCurrency = args["newCurrency"] as! String
                self.setCurrency(newCurrency: newCurrency)
            case "setShippingAmount" :
                guard let args = call.arguments as? [String : Any] else {return}
                let amount = args["amount"] as! Double
                self.setShippingAmount(amount: amount)
            case "setDiscount" :
                guard let args = call.arguments as? [String : Any] else {return}
                let amount = args["amount"] as! Double
                let name = args["name"] as! String
                self.setDiscount(amount: amount, name: name)
            case "paymentOrder" :
                self.paymentOrder(fResult: result)
            case "clearItem" :
                self.clearItem()
            case "authoriseOrder" :
                guard let args = call.arguments as? [String : Any] else {return}
                let orderId = args["orderId"] as! String
                self.authoriseOrder(orderId: orderId, fResult: result)
            case "refunds" :
                guard let args = call.arguments as? [String : Any] else {return}
                let orderId = args["orderId"] as! String
                let jsonData = args["refunds"] as! String
                self.refunds(orderId: orderId, jsonData: jsonData, fResult: result)
            case "capturePayment" :
                guard let args = call.arguments as? [String : Any] else {return}
                let jsonData = args["capturePayment"] as! String
                self.capturePayment(jsonData: jsonData, fResult: result)
            case "orderDetail" :
                guard let args = call.arguments as? [String : Any] else {return}
                let orderId = args["orderId"] as! String
                self.orderDetail(orderId: orderId, fResult: result)
            case "cancelOrder" :
                guard let args = call.arguments as? [String : Any] else {return}
                let orderId = args["orderId"] as! String
                let jsonData = args["cancelOrder"] as! String
                self.cancelOrder(orderId: orderId, jsonData: jsonData, fResult: result)
            case "updateOrderReference" :
                guard let args = call.arguments as? [String : Any] else {return}
                let orderId = args["orderId"] as! String
                let orderReferenceId = args["orderReferenceId"] as! String
                self.updateOrderReference(orderId: orderId, orderReferenceId: orderReferenceId, fResult: result)
            case "renderCartPage" :
                guard let args = call.arguments as? [String : Any] else {return}
                let language = args["language"] as! String
                let country = args["country"] as! String
                let publicKey = args["publicKey"] as! String
                let amount = args["amount"] as! Double
                self.renderWidgetCartPage(language: language, country: country, publicKey: publicKey, amount: amount, result: result)
            case "renderProduct" :
                guard let args = call.arguments as? [String : Any] else {return}
                let language = args["language"] as! String
                let country = args["country"] as! String
                let publicKey = args["publicKey"] as! String
                let amount = args["amount"] as! Double
                self.renderWidgetProduct(language: language, country: country, publicKey: publicKey, amount: amount, result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
    }
    
    func initialize(token: String, apiUrl: String, pushUrl: String, publishKey: String, notificationToken: String, isSandbox: Bool) {
        TamaraSDKPayment.shared.initialize(token: token, apiUrl: apiUrl, pushUrl: pushUrl, publishKey: publishKey, notificationToken: notificationToken, isSandbox: isSandbox)
    }
    
    func createOrder(orderReferenceId: String, description: String) {
        TamaraSDKPayment.shared.createOrder(orderReferenceId: orderReferenceId, description: description)
    }
    
    func setCountry(countryCode: String, currency: String) {
        TamaraSDKPayment.shared.setCountry(countryCode: countryCode, currency: currency)
    }
    
    func setPaymentType(paymentType: String) {
        TamaraSDKPayment.shared.setPaymentType(paymentType: paymentType)
    }

    func setInstalments(instalments: Int) {
        TamaraSDKPayment.shared.setInstalments(instalments: instalments)
    }
    
    func setCustomerInfo(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        email: String,
        isFirstOrder: Bool
    ) {
        TamaraSDKPayment.shared.setCustomerInfo(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, email: email, isFirstOrder: isFirstOrder)
    }
    
    func addItem(
        name: String,
        referenceId: String,
        sku: String,
        type: String,
        unitPrice: Double,
        tax: Double,
        discount: Double,
        quantity: Int
    ) {
        TamaraSDKPayment.shared.addItem(name: name, referenceId: referenceId, sku: sku, type: type, unitPrice: unitPrice, tax: tax, discount: discount, quantity: quantity)
    }
    
    func setShippingAddress(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        addressLine1: String,
        addressLine2: String,
        country: String,
        region: String,
        city: String
    ) {
        TamaraSDKPayment.shared.setShippingAddress(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, addressLine1: addressLine1, addressLine2: addressLine2, country: country, region: region, city: city)
    }
    
    func setBillingAddress(
        firstName: String,
        lastName: String,
        phoneNumber: String,
        addressLine1: String,
        addressLine2: String,
        country: String,
        region: String,
        city: String
    ) {
        TamaraSDKPayment.shared.setBillingAddress(firstName: firstName, lastName: lastName, phoneNumber: phoneNumber, addressLine1: addressLine1, addressLine2: addressLine2, country: country, region: region, city: city)
    }
    
    func setCurrency(newCurrency: String) {
        TamaraSDKPayment.shared.setCurrency(newCurrency: newCurrency)
    }
    
    func setShippingAmount(amount: Double) {
        TamaraSDKPayment.shared.setShippingAmount(amount: amount)
    }
    
    func setDiscount(amount: Double, name: String) {
        TamaraSDKPayment.shared.setDiscount(amount: amount, name: name)
    }
    
    func paymentOrder(fResult: @escaping FlutterResult) {
        TamaraSDKPayment.shared.startPayment() { result in
            switch result {
            case let .success(response):
                fResult(response.convertToJson())
            case let .failure(error):
                fResult(FlutterError(code: "",
                                    message: error.localizedDescription,
                                        details: nil))
            }
        }

    }
    
    func clearItem() {
        TamaraSDKPayment.shared.clearItem()
    }
    
    func authoriseOrder(orderId: String, fResult: @escaping FlutterResult){
        TamaraSDKPayment.shared.authoriseOrder(orderId: orderId) { result in
            switch result {
            case let .success(data):
                fResult(data.convertToJson())
            case let .failure(error):
                fResult(error.localizedDescription)
            }
        }
    }
    
    func refunds(orderId: String, jsonData: String, fResult: @escaping FlutterResult){
        TamaraSDKPayment.shared.refunds(orderId: orderId, jsonData: jsonData) { result in
            switch result {
            case let .success(data):
                fResult(data.convertToJson())
            case let .failure(error):
                fResult(error.localizedDescription)
            }
        }
    }
    
    func orderDetail(orderId: String, fResult: @escaping FlutterResult) {
        TamaraSDKPayment.shared.orderDetail(orderId: orderId) { result in
            switch result {
            case let .success(response):
                fResult(response.convertToJson())
            case let .failure(error):
                fResult(FlutterError(code: "",
                                    message: error.localizedDescription,
                                        details: nil))
            }
        }
    }
    
    func capturePayment(jsonData: String, fResult: @escaping FlutterResult){
        TamaraSDKPayment.shared.capturePayment(jsonData: jsonData) { result in
            switch result {
            case let .success(response):
                fResult(response.convertToJson())
            case let .failure(error):
                fResult(error.localizedDescription)
            }
        }
    }
    
    func cancelOrder(orderId: String, jsonData: String, fResult: @escaping FlutterResult){
        //get order detail -> get total amount
        TamaraSDKPayment.shared.orderDetail(orderId: orderId) { result in
            switch result {
            case let .success(response):
                TamaraSDKPayment.shared.cancelOrder(orderId: orderId, jsonData: CancelOrder(discountAmount: nil, items: Array(), shippingAmount: nil, taxAmount: nil, totalAmount: response.totalAmount!).convertToJson()) { result in
                    switch result {
                    case let .success(data):
                        fResult(data.convertToJson())
                    case let .failure(error):
                        fResult(error.localizedDescription)
                    }
                }
            case let .failure(error):
                fResult(error.localizedDescription)
            }
        }
    }
    
    func updateOrderReference(orderId: String, orderReferenceId: String, fResult: @escaping FlutterResult) {
        TamaraSDKPayment.shared.updateOrderReference(orderId: orderId, orderReferenceId: orderReferenceId) { result in
            switch result {
            case let .success(data):
                fResult(data.convertToJson())
            case let .failure(error):
                fResult(error.localizedDescription)
            }
        }
    }
    
    func renderWidgetCartPage(language: String, country: String, publicKey: String,
                              amount: Double, result: @escaping FlutterResult) {
        TamaraSDKPayment.shared.renderWidgetCartPage(language: language, country: country, publicKey: publicKey, amount: amount){ response, error, code in
            if let response = response as? NSDictionary {
                switch code {
                case 200:
                    if let theJSONData = try? JSONSerialization.data(
                        withJSONObject: response,
                        options: []) {
                        let data = WidgetProperties(jsonData: theJSONData)
                        result(data.convertToJson())
                    }
                default:
                    result(response.allValues)
                }
            }
       }
    }
    
    func renderWidgetProduct(language: String, country: String, publicKey: String,
                              amount: Double, result: @escaping FlutterResult) {
        TamaraSDKPayment.shared.renderWidgetProduct(language: language, country: country, publicKey: publicKey, amount: amount){ response, error, code in
            if let response = response as? NSDictionary {
                switch code {
                case 200:
                    if let theJSONData = try? JSONSerialization.data(
                        withJSONObject: response,
                        options: []) {
                        let data = WidgetProperties(jsonData: theJSONData)
                        result(data.convertToJson())
                    }
                default:
                    result(response.allValues)
                }
            }
       }
    }
}
